kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/FindNodePendingRequest.java

```diff
@@ -0,0 +1,272 @@
+package ingsw.group1.kademlia.pendingrequests;
+
+import androidx.annotation.NonNull;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+
+import ingsw.group1.kademlia.ActionPropagator;
+import ingsw.group1.kademlia.BinarySet;
+import ingsw.group1.kademlia.KadAction;
+import ingsw.group1.kademlia.KadActionsBuilder;
+import ingsw.group1.kademlia.NodeDataProvider;
+import ingsw.group1.kademlia.NodeUtils;
+import ingsw.group1.kademlia.PeerNode;
+import ingsw.group1.kademlia.listeners.FindNodeResultListener;
+import ingsw.group1.msglibrary.SMSPeer;
+
+/**
+ * Class defining an implementation of {@link PendingRequest} for a FIND_NODE type Request, as
+ * defined in the Kademlia protocol.
+ * After completion the {@code PendingRequest} should not receive further calls to
+ * {@link PendingRequest#nextStep(KadAction)}.
+ *
+ * @author Riccardo De Zen
+ * CODE REVIEW
+ */
+public class FindNodePendingRequest implements PendingRequest {
+
+    private static KadActionsBuilder actionBuilder = new KadActionsBuilder();
+
+    //These constants are soon to be moved to an appropriate class.
+    /**
+     * The number of Nodes contacted on each iteration and the number of Nodes every contacted
+     * Node will send back.
+     */
+    private static final int KAD_K = 10;
+    /**
+     * The length of IDs in the Network and the number of buckets in the Routing Table.
+     */
+    private static final int KAD_N = 128;
+
+    private RequestState requestState = RequestState.IDLE;
+    private int totalStepsTaken = 0;
+
+    private int operationId;
+    private BinarySet targetId;
+    private ActionPropagator actionPropagator;
+    private NodeDataProvider<BinarySet, PeerNode> nodeProvider;
+    private FindNodeResultListener resultListener;
+
+    //Map used to keep track of all visited Nodes
+    private TreeMap<BinarySet, PeerNode> visitedNodes = new TreeMap<>();
+
+    //Set used to keep track of the Peers from which we are still waiting for a Response.
+    private Set<PeerNode> pendingResponses = new TreeSet<>();
+
+    //Set used to keep track of the Peers we received as Responses from the Peers we contacted.
+    private Set<PeerNode> peerBuffer = new TreeSet<>();
+
+    /**
+     * The number of Responses we expect. It increases every time a Node responds for the first
+     * time and it decreases by one for every received Response.
+     */
+    private int expectedResponses = 0;
+
+    /**
+     * Default constructor.
+     *
+     * @param operationId      the unique id for this PendingRequest operation.
+     * @param targetId         the id of the Node we are looking for.
+     * @param actionPropagator a valid {@link ActionPropagator}.
+     * @param nodeProvider     a valid {@link NodeDataProvider}.
+     * @param resultListener   a valid listener to this {@code PendingRequest}'s Result.
+     */
```

Comment:
I agree 5 parameters are way too many, there used to be a Factory class for PendingRequest, which reduced the number of required to two, but I had to get rid of it because of trouble with initializing some types of Requests. The removal of the factory is also the same reason for the PendingRequestManager having 5 methods with almost identical bodies, he kinda took the place of the Factory. I'll definitely be reducing the parameters in the near future, when we will implement some Object Pools for most classes in the network